### PR TITLE
Upgrade all JavaScript dependencies

### PR DESCRIPTION
**Pull request by nwalters512**

### CODE REVIEW

File: apps/prairielearn/elements/pl-order-blocks/dag_checker.py
Comment (nwalters512) [inline apps/prairielearn/elements/pl-order-blocks/dag_checker.py:1]: @eliotwrobson looks like Pyright got a little smarter!

**Review by SethPoulsen — APPROVED** [review]

**Comment by SethPoulsen** [comment]
Looks good, always happy to see the code that I own get shorter :)

File: apps/prairielearn/package.json [hunk 1]
```diff
@@ -78,37 +78,37 @@
...
-    "d3": "^7.8.5",
+    "d3": "^7.9.0",
     "date-fns": "^2.30.0",
-    "date-fns-tz": "^2.0.0",
+    "date-fns-tz": "^2.0.1",
     "debug": "^4.3.4",
     "delegated-events": "^1.1.2",
-    "dockerode": "^3.3.5",
-    "dompurify": "^3.0.9",
+    "dockerode": "^4.0.2",
```
Comment (mwest1066) [inline apps/prairielearn/package.json:86]: I couldn't figure out what the breaking changes were here, but I did see the code change when calling `docker.listContainers()`.

**Review by mwest1066 — APPROVED** [review]

File: apps/prairielearn/package.json (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/package.json:86]: Follow breadcrumbs from here: https://github.com/apocas/dockerode/issues/751

As for the `listContainers` change, that was a genuine bug uncovered by the introduction of better types!

**Review by jonatanschroeder — APPROVED** [review]

File: apps/prairielearn/package.json (hunk 1 above)
Comment (mwest1066) [reply apps/prairielearn/package.json:86]: Oh nice! 🙏 for fixing this bug upstream. And yay for types!

**Pull request merged**